src/seq.rs

```diff
@@ -110,116 +117,204 @@ pub fn sample_slice<R, T>(rng: &mut R, slice: &[T], amount: usize) -> Vec<T>
+/// otherwise the values may only partially shuffled, depending on the
+/// algorithm used (i.e. biases may exist in the ordering of sampled elements).
+/// Depending on the algorithm used internally, full shuffling may add
+/// significant overhead for `amount` > 10 or so, but not more than double
+/// the time and often much less.
 ///
-/// This implementation uses `O(amount)` time and memory.
+/// This method is used internally by the slice sampling methods, but it can
+/// sometimes be useful to have the indices themselves so this is provided as
+/// an alternative.
 ///
-/// This method is used internally by the slice sampling methods, but it can sometimes be useful to
-/// have the indices themselves so this is provided as an alternative.
+/// The implementation used is not specified; we automatically select the
+/// fastest available implementation for the `length` and `amount` parameters
+/// (based on detailed profiling on an Intel Haswell CPU). Roughly speaking,
+/// complexity is `O(amount)`, except that when `amount` is small, performance
+/// is closer to `O(amount^2)`, and when `length` is close to `amount` then
+/// `O(length)`.
 ///
-/// Panics if `amount > length`
-pub fn sample_indices<R>(rng: &mut R, length: usize, amount: usize) -> Vec<usize>
+/// Note that we only support `u32` indices since this covers the vast majority
+/// of uses, and performance is significantly better than with `u64`.
+/// 
+/// If an allocation-free `no_std` function is required, it is suggested
+/// to adapt the internal `sample_indices_floyd` implementation.
+///
+/// Panics if `amount > length` or if `length` is not reprentable as a `u32`.
+pub fn sample_indices<R>(rng: &mut R, length: usize, amount: usize,
+    shuffled: bool) -> Vec<u32>
     where R: Rng + ?Sized,
 {
     if amount > length {
-        panic!("`amount` must be less than or equal to `slice.len()`");
+        panic!("`amount` of samples must be less than or equal to `length`");
     }
-
-    // We are going to have to allocate at least `amount` for the output no matter what. However,
-    // if we use the `cached` version we will have to allocate `amount` as a HashMap as well since
-    // it inserts an element for every loop.
-    //
-    // Therefore, if `amount >= length / 2` then inplace will be both faster and use less memory.
-    // In fact, benchmarks show the inplace version is faster for length up to about 20 times
-    // faster than amount.
-    //
-    // TODO: there is probably even more fine-tuning that can be done here since
-    // `HashMap::with_capacity(amount)` probably allocates more than `amount` in practice,
-    // and a trade off could probably be made between memory/cpu, since hashmap operations
-    // are slower than array index swapping.
-    if amount >= length / 20 {
-        sample_indices_inplace(rng, length, amount)
+    if length > (::core::u32::MAX as usize) {
+        panic!("`length` is not representable as `u32`");
+    }
```

Comment:
I am really not a fan of this change. All algorithms can work with similar performance as when using an `usize`, but it would take a little extra complexity. It would simply make things work without special cases.

I gave it a try for Floyds algorithm [here](https://github.com/pitdicker/rand/commit/dd38432c3317c040688464fb914040a5d302a123#diff-f0a22d0ca5f8f67f38198c46ef26ad68R208).

And played with optimizing a partial shuffle in https://github.com/pitdicker/rand/blob/slice_random/src/seq.rs#L107 (used in `sample_indices_inplace`).I am sure something we can come up with something cleaner than that that also works with `sample_indices_cache`.

Reply:
That will still slow down Floyd's algorithm with small indices since the output vector is larger, hence searching it (`contains`) is slower when more memory needs to be read. In particular, more results can fit in the cache with smaller indices.

This means that it can be faster to convert the output back to `usize` later, although that's ugly. Only outputting `Vec<u32>` was motivated by this problem.